File: tests/algorithms/x/Java/dynamic_programming/palindrome_partitioning.java
public class Main {

    static int min_partitions(String s) {
        int n = _runeLen(s);
        int[] cut = ((int[])(new int[]{}));
        int i = 0;
        while (i < n) {
            cut = ((int[])(java.util.stream.IntStream.concat(java.util.Arrays.stream(cut), java.util.stream.IntStream.of(0)).toArray()));
            i = i + 1;
        }
        boolean[][] pal = ((boolean[][])(new boolean[][]{}));
        i = 0;
        while (i < n) {
            boolean[] row = ((boolean[])(new boolean[]{}));
            int j = 0;
            while (j < n) {
                row = ((boolean[])(appendBool(row, false)));
                j = j + 1;
            }
            pal = ((boolean[][])(appendObj(pal, row)));
            i = i + 1;
        }
        i = 0;
        while (i < n) {
            int mincut = i;
            int j_1 = 0;
            while (j_1 <= i) {
                if ((s.substring(i, i+1).equals(s.substring(j_1, j_1+1))) && (i - j_1 < 2 || ((Boolean)(pal[j_1 + 1][i - 1])))) {
pal[j_1][i] = true;
                    if (j_1 == 0) {
                        mincut = 0;
                    } else {
                        int candidate = cut[j_1 - 1] + 1;
                        if (candidate < mincut) {
                            mincut = candidate;
                        }
                    }
                }
                j_1 = j_1 + 1;
            }
cut[i] = mincut;
            i = i + 1;
        }
        return cut[n - 1];
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            System.out.println(min_partitions("aab"));
            System.out.println(min_partitions("aaa"));
            System.out.println(min_partitions("ababbbabbababa"));
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static boolean[] appendBool(boolean[] arr, boolean v) {
        boolean[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }

    static <T> T[] appendObj(T[] arr, T v) {
        T[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }

    static int _runeLen(String s) {
        return s.codePointCount(0, s.length());
    }
}

File: tests/algorithms/x/Pascal/dynamic_programming/palindrome_partitioning.pas
{$mode objfpc}
program Main;
uses SysUtils;
type IntArray = array of integer;
type BoolArray = array of boolean;
type BoolArrayArray = array of BoolArray;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  s: string;
function min_partitions(s: string): integer; forward;
function min_partitions(s: string): integer;
var
  min_partitions_n: integer;
  min_partitions_cut: array of integer;
  min_partitions_i: integer;
  min_partitions_pal: array of BoolArray;
  min_partitions_row: array of boolean;
  min_partitions_j: integer;
  min_partitions_mincut: integer;
  min_partitions_candidate: integer;
begin
  min_partitions_n := Length(s);
  min_partitions_cut := [];
  min_partitions_i := 0;
  while min_partitions_i < min_partitions_n do begin
  min_partitions_cut := concat(min_partitions_cut, IntArray([0]));
  min_partitions_i := min_partitions_i + 1;
end;
  min_partitions_pal := [];
  min_partitions_i := 0;
  while min_partitions_i < min_partitions_n do begin
  min_partitions_row := [];
  min_partitions_j := 0;
  while min_partitions_j < min_partitions_n do begin
  min_partitions_row := concat(min_partitions_row, [false]);
  min_partitions_j := min_partitions_j + 1;
end;
  min_partitions_pal := concat(min_partitions_pal, [min_partitions_row]);
  min_partitions_i := min_partitions_i + 1;
end;
  min_partitions_i := 0;
  while min_partitions_i < min_partitions_n do begin
  min_partitions_mincut := min_partitions_i;
  min_partitions_j := 0;
  while min_partitions_j <= min_partitions_i do begin
  if (s[min_partitions_i+1] = s[min_partitions_j+1]) and (((min_partitions_i - min_partitions_j) < 2) or min_partitions_pal[min_partitions_j + 1][min_partitions_i - 1]) then begin
  min_partitions_pal[min_partitions_j][min_partitions_i] := true;
  if min_partitions_j = 0 then begin
  min_partitions_mincut := 0;
end else begin
  min_partitions_candidate := min_partitions_cut[min_partitions_j - 1] + 1;
  if min_partitions_candidate < min_partitions_mincut then begin
  min_partitions_mincut := min_partitions_candidate;
end;
end;
end;
  min_partitions_j := min_partitions_j + 1;
end;
  min_partitions_cut[min_partitions_i] := min_partitions_mincut;
  min_partitions_i := min_partitions_i + 1;
end;
  exit(min_partitions_cut[min_partitions_n - 1]);
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  writeln(min_partitions('aab'));
  writeln(min_partitions('aaa'));
  writeln(min_partitions('ababbbabbababa'));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
